File: 22. Ch-5-5.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.flag = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.flag = True
        print(f"Inserted: {word}")

    def search(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                return False
            node = node.children[ch]
        if node.flag:
            return True
        else:
            return False

    def remove(self, word):
        if not self.search(word):
            print(f"Not found: {word}")
            return
        self.removeRecursion(self.root, word, 0)
        print(f"Removed: {word}")

    def removeRecursion(self, node, word, index):
        if index == len(word):
            node.flag = False
            return len(node.children) == 0
        ch = word[index]
        deleteChild = self.removeRecursion(node.children[ch], word, index + 1)

        if deleteChild:
            del node.children[ch]
            return not node.flag and len(node.children) == 0
        return False

# -------- MAIN --------
trie = Trie()

while True:
    op = int(input())
    if op == 1:
        word = input()
        trie.insert(word)
    elif op == 2:
        word = input()
        res=trie.search(word)
        if res:
            print(f"Found: {word}")
        else:
            print(f"Not found: {word}")
    elif op == 3:
        word = input()
        trie.remove(word)
    elif op == 4:
        break

File: 22. Ch-5-5.java
import java.util.*;

class TrieNode 
{
    Map<Character, TrieNode> children;
    boolean flag;
    public TrieNode() 
    {
        children = new HashMap<>();
        flag = false;
    }
}

class Trie 
{
    TrieNode root;
    Trie() 
    {
        root = new TrieNode();
    }
    void insert(String word) 
    {
        TrieNode current = root;
        for (int i = 0; i < word.length(); i++) 
        {
            char ch = word.charAt(i);
            TrieNode node = current.children.get(ch);
            if (node == null) {
                node = new TrieNode();
                current.children.put(ch, node);
            }
            current = node;
        }
        current.flag = true;
        System.out.println("Inserted: " + word);
    }

    boolean search(String word) 
    {
        TrieNode current = root;
        for (int i = 0; i < word.length(); i++) 
        {
            char ch = word.charAt(i);
            TrieNode node = current.children.get(ch);
            if (node == null) 
            {
                return false;
            }
            current = node;
        }
        return current.flag;
    }

    void remove(String word) 
    {
        if (!search(word)) 
        {
            System.out.println("Not found: " + word);
            return;
        }
        removeRecursion(root, word, 0);
        System.out.println("Removed: " + word);
    }

    public boolean removeRecursion(TrieNode current, String word, int index) 
    {
        if (index == word.length()) 
        {
            current.flag = false;
            return current.children.size() == 0;
        }

        char ch = word.charAt(index);
        TrieNode node = current.children.get(ch);
        boolean shouldDeleteChild = removeRecursion(node, word, index + 1);

        if (shouldDeleteChild) 
        {
            current.children.remove(ch);
            return !current.flag && current.children.size() == 0;
        }
        return false;
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        Trie trie = new Trie();
        while (true) 
        { 
            int op = Integer.valueOf(sc.nextLine().trim());
            if (op == 1) 
            {
                trie.insert(sc.nextLine());
            } 
            else if (op == 2) 
            {
                String word = sc.nextLine();
                if (trie.search(word)) 
                {
                    System.out.println("Found: " + word);
                } 
                else 
                {
                    System.out.println("Not found: " + word);
                }
            } 
            else if (op == 3) 
            {
                trie.remove(sc.nextLine());
            } 
            else if (op == 4) 
            {
                break;
            }
        }
    }
}
